spec/services/openid_connect_redirector_spec.rb

```diff
@@ -0,0 +1,91 @@
+require 'rails_helper'
+
+RSpec.describe OpenidConnectRedirector do
+  include Rails.application.routes.url_helpers
+
+  let(:redirect_uri) { 'http://localhost:7654/' }
+  let(:state) { SecureRandom.hex }
+  let(:service_provider) { ServiceProvider.from_issuer('urn:gov:gsa:openidconnect:sp:server') }
+  let(:errors) { ActiveModel::Errors.new(nil) }
+
+  describe '.from_request_url' do
+    it 'builds a redirector from an OpenID request_url' do
+      request_url = openid_connect_authorize_url(
+        client_id: service_provider.issuer,
+        redirect_uri: redirect_uri,
+        state: state
+      )
+
+      result = OpenidConnectRedirector.from_request_url(request_url)
+
+      expect(result).to be_a(OpenidConnectRedirector)
+      expect(result.send(:redirect_uri)).to eq(redirect_uri)
+      expect(result.send(:service_provider)).to eq(service_provider)
+      expect(result.send(:state)).to eq(state)
+    end
+  end
+
+  subject(:redirector) do
+    OpenidConnectRedirector.new(
+      redirect_uri: redirect_uri,
+      service_provider: service_provider,
+      state: state,
+      errors: errors
+    )
+  end
+
+  describe '#validate' do
+    context 'with a valid redirect_uri' do
+      let(:redirect_uri) { 'http://localhost:7654/result/more/extra' }
+      it 'is valid' do
+        redirector.validate
+        expect(errors).to be_empty
+      end
+    end
+
+    context 'with a malformed redirect_uri' do
+      let(:redirect_uri) { ':aaaa' }
+      it 'has errors' do
+        redirector.validate
+        expect(errors[:redirect_uri]).
+          to include(t('openid_connect.authorization.errors.redirect_uri_invalid'))
+      end
+    end
+
+    context 'with a redirect_uri not registered to the client' do
```

Comment:
I'll rename the test to reference the service provider. Since our class is named `ServiceProvider` I think that is least likely to change any time soon